### DISCUSSION TITLE
How can I add permission to users by default at their registration time?

### DISCUSSION THREAD

**Discussion by sturlath** [abpframework/abp#9428] [Q&A] [answered]
Hi

I want to give all (self) registered user a default set of permissions/access. That permissions will be used to show e.g. different menus that Tenant/Host see.

So the question is : **how can I  hook into the "user was created" event and add the permission to the user?** 
Or is there some better way?

**Comment by realLiangshiwei** [comment]
https://docs.abp.io/en/abp/latest/Local-Event-Bus#events-with-past-tense

You can subscribe to user-created domain events

**Reply by sturlath** [reply]
I had tried that but I didn´t step into that code. Could that be because I'm using the Volo.Account.Pro source code? Do I maybe need to add this code under that namespace? Where?

**Reply by sturlath** [reply]
Ok ok my bad! I had the identity service starting up with no debug.. the event is getting hit (when placed in the Account source module, not in my Domain project)..
I will now be [adding permissions]( https://github.com/abpframework/abp/issues/8600#issuecomment-819128957 )

**Reply by sturlath** [reply]
Just for others, you can't add the *.Application.Contracts project (where the permissions are located) to the source module because of compatibility issue between the projects (these projects will hopefully be updated to .net5) so I´ll just add the permission class to a new separate project and add it to the source module and my Contracts project.

`MyApp.Domain is not compatible with netstandard2.0 (.NETStandard,Version=v2.0). Project MyApp.Domain supports: net5.0 (.NETCoreApp,Version=v5.0)	Volo.Abp.Account.Pro.Shared.Application	C:\Dev\aspnet-core\modules\Volo.Account.Pro\src\Volo.Abp.Account.Pro.Shared.Application\Volo.Abp.Account.Pro.Shared.Application.csproj`

**Reply by sturlath** [reply]
Ok I have been at this for some time and I just can´t get it to work! I only get `Undefined permission` error for my permissions.

I'm able to set these permissions through the UI for these users no problem and explore them in the db so I know (famous last words) I'm passing the correct permission values down.  I even try to add `AbpAccount.SettingManagement` since that is in the module but that doesn't matter.

So what am I missing here @realLiangshiwei (or somebody else of course)?

```csharp
using System.Threading.Tasks;
using Volo.Abp.DependencyInjection;
using Volo.Abp.Domain.Entities.Events;
using Volo.Abp.EventBus;
using Volo.Abp.Identity;
using Volo.Abp.PermissionManagement;

namespace Volo.Abp.Account.Pro.Public.Application
{
    public class UserHandler : ILocalEventHandler<EntityCreatedEventData<IdentityUser>>, ITransientDependency
    {
        private readonly IPermissionManager permissionManager;
        public UserHandler(IPermissionManager permissionManager)
        {
            this.permissionManager = permissionManager;
        }

        public async Task HandleEventAsync(EntityCreatedEventData<IdentityUser> eventData)
        {
            try
            {
                await permissionManager.SetForUserAsync(eventData.Entity.Id, "MyProject.MyAction.Default", true);
                //Same with this one
                await permissionManager.SetAsync("MyProject.MyAction.Default", "U", eventData.Entity.Id.ToString(), true);
            }
            catch (System.Exception ex)
            {
                // I always get this error
                //Undefined permission: MyProject.MyAction.Default
                //Volo.Abp.AbpException: Undefined permission: MyProject.MyAction.Default
                //   at Volo.Abp.Authorization.Permissions.PermissionDefinitionManager.Get(String name)
                //   at Volo.Abp.PermissionManagement.PermissionManager.SetAsync(String permissionName, String providerName, String providerKey, Boolean isGranted)
                //   at Volo.Abp.Account.Pro.Public.Application.UserHandler.HandleEventAsync(EntityCreatedEventData`1 eventData) in 
                //    C:\Dev\MyProject\aspnet - core\modules\Volo.Account.Pro\src\Volo.Abp.Account.Pro.Public.Application\Volo\Abp\Account\UserHandler.cs:line 26
                //   at Volo.Abp.EventBus.EventBusBase.TriggerHandlerAsync(IEventHandlerFactory asyncHandlerFactory, Type eventType, Object eventData, List`1 exceptions)
                //   at Volo.Abp.EventBus.Local.LocalEventBus.PublishAsync(Type eventType, Object eventData)
                //   at Volo.Abp.Domain.Entities.Events.EntityChangeEventHelper.<> c__DisplayClass34_0.<< TriggerEventWithEntity > b__0 > d.MoveNext()
            }
        }
    }
}
```

**Reply by sturlath** [reply]
Ok I think I might have found the reason (not the cure)! 

If I try to get back the newly created user it only returns `AbpIdenity.*` permissions (all of course with IsGranted= false) . 

![image](https://user-images.githubusercontent.com/2386572/123454500-296f3580-d5d0-11eb-99ee-938063780188.png)

And trying this, it works `await permissionManager.SetForUserAsync(eventData.Entity.Id, "AbpIdentity.Roles.Update", true);`
How do I "plug" my common permissions into the identity module?

**Reply by sturlath** [reply]
I have been trying to figure this out @realLiangshiwei and now believe I need to register my first `PermissionManagementProvider ` like [here ](https://docs.abp.io/en/abp/latest/Modules/Permission-Management#permission-management-providers). 

Am I on the right track?

So add this code somewhere in my code 
```csharp
Configure<PermissionManagementOptions>(options` =>
{
    options.ManagementProviders.Add<MyCustomPermissionManagementProvider>();
});
```

I really don't see that I´m doing something wrong and can´t believe that I'm the only one that is trying to add permissions to a user using a Account module. I´ll do a PR to the docks after I get help with this..

**Reply by realLiangshiwei** [reply]
Hi,

I don't know why you need to this, can you use CLI to create a simple project to reproduce?

**Reply by sturlath** [reply]
I’ll create a project for you np but until then here is some explanation on what I'm doing.

This is a SaaS marketplace (not sure if that’s the right word). Concert holders create a Tenant and create (none-specific-Tenant) events. Ordinary (un-logged in) users can browse list of events from every Tenant in the system and sign up (and pay for) to those events.

Tenants and Ordinary users have different access (different menus). The Tenant needs to manage events and see graphs etc but the Ordinary user see his/her upcoming payed events and payment history/info etc.(and lot more).

And then its me the Host that will be managing the system.

So when an Ordinary user registers (after e.g. clicking on an event) that he/she wants to attend I want to distinguish them from a Tenant or a Host. That's why I’m **trying to add permissions on after-created.**

**Is there another way I can accomplish this or am I trying to use the Tenant system incorrectly?**

If you still need a CLI project  let me know and Ill email it to you later.

**Reply by sturlath** [reply]
For (circular) reference I updated [an old support question](https://support.abp.io/QA/Questions/1084/Architectural-question-on-whether-to-choose-Multi-Tenancy-or-not) I had (3 months ago). It feels like this belongs there (since it got so complicated and not just off the bat `SetForUserAsync()` call..

**Reply by sturlath** [reply]
@realLiangshiwei? Do you still need a project or is isnt it clear what I’m trying to do? Is this not a known use case? It must be…right?

**Reply by realLiangshiwei** [reply]
Yes I need a project to understand your problem

**Reply by sturlath** [reply]
You got mail!

**Comment by sturlath** [comment, marked as answer]
We got this solved and it was related to me using a separate identity server. I created a [docs PR](https://github.com/abpframework/abp/pull/9485) that you can view to better understand this. Hopefully this update prevents somebody else going through my pain also ;-)
